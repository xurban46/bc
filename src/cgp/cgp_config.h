/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

// hard configuration - affects compilation (array sizes etc.)
#define CGP_COLS 8
#define CGP_ROWS 4
#define CGP_INPUTS 9
#define CGP_OUTPUTS 1
#define CGP_LBACK 1
