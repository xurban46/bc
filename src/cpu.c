/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


/*
    CPU new instruction set check.
    Source: https://software.intel.com/en-us/articles/how-to-detect-new-instruction-support-in-the-4th-generation-intel-core-processor-family
 */


#include "cpu.h"


#if defined(__INTEL_COMPILER) && (__INTEL_COMPILER >= 1300)

#include <immintrin.h>

int check_4th_gen_intel_core_features()
{
    const int the_4th_gen_features =
        (_FEATURE_AVX2 | _FEATURE_FMA | _FEATURE_BMI | _FEATURE_LZCNT | _FEATURE_MOVBE);
    return _may_i_use_cpu_feature( the_4th_gen_features );
}

int check_sse4_1()
{
    return _may_i_use_cpu_feature(_FEATURE_SSE4_1);
}

int check_sse2()
{
    return _may_i_use_cpu_feature(_FEATURE_SSE2);
}

#else /* non-Intel compiler */

#include <stdint.h>
#if defined(_MSC_VER)
# include <intrin.h>
#endif

void run_cpuid(uint32_t eax, uint32_t ecx, uint32_t* abcd)
{
    #if defined(_MSC_VER)
        __cpuidex(abcd, eax, ecx);
    #else
        uint32_t ebx = 0, edx;
    # if defined( __i386__ ) && defined ( __PIC__ )
         /* in case of PIC under 32-bit EBX cannot be clobbered */
        __asm__ ( "movl %%ebx, %%edi \n\t cpuid \n\t xchgl %%ebx, %%edi" : "=D" (ebx),
    # else
        #pragma GCC diagnostic ignored "-Wuninitialized"
        __asm__ ( "cpuid" : "+b" (ebx),
    # endif
                  "+a" (eax), "+c" (ecx), "=d" (edx) );
        abcd[0] = eax; abcd[1] = ebx; abcd[2] = ecx; abcd[3] = edx;
    #endif
}

int check_xcr0_ymm()
{
        uint32_t xcr0;
    #if defined(_MSC_VER)
        xcr0 = (uint32_t)_xgetbv(0);  /* min VS2010 SP1 compiler is required */
    #else
        __asm__ ("xgetbv" : "=a" (xcr0) : "c" (0) : "%edx" );
    #endif
        return ((xcr0 & 6) == 6); /* checking if xmm and ymm state are enabled in XCR0 */
}


int check_4th_gen_intel_core_features()
{
    uint32_t abcd[4];
    uint32_t fma_movbe_osxsave_mask = ((1 << 12) | (1 << 22) | (1 << 27));
    uint32_t avx2_bmi12_mask = (1 << 5) | (1 << 3) | (1 << 8);

    /* CPUID.(EAX=01H, ECX=0H):ECX.FMA[bit 12]==1   &&
       CPUID.(EAX=01H, ECX=0H):ECX.MOVBE[bit 22]==1 &&
       CPUID.(EAX=01H, ECX=0H):ECX.OSXSAVE[bit 27]==1 */
    run_cpuid( 1, 0, abcd );
    if ( (abcd[2] & fma_movbe_osxsave_mask) != fma_movbe_osxsave_mask )
        return 0;

    if ( ! check_xcr0_ymm() )
        return 0;

    /*  CPUID.(EAX=07H, ECX=0H):EBX.AVX2[bit 5]==1  &&
        CPUID.(EAX=07H, ECX=0H):EBX.BMI1[bit 3]==1  &&
        CPUID.(EAX=07H, ECX=0H):EBX.BMI2[bit 8]==1  */
    run_cpuid( 7, 0, abcd );
    if ( (abcd[1] & avx2_bmi12_mask) != avx2_bmi12_mask )
        return 0;

    /* CPUID.(EAX=80000001H):ECX.LZCNT[bit 5]==1 */
    run_cpuid( 0x80000001, 0, abcd );
    if ( (abcd[2] & (1 << 5)) == 0)
        return 0;

    return 1;
}

/**
 * Checks whether current CPU supports SSE4.1 instruction set
 *
 * Source:
 *     http://stackoverflow.com/questions/6121792/how-to-check-if-a-cpu-supports-the-sse3-instruction-set
 *     http://stackoverflow.com/a/7495023
 */
int check_sse4_1()
{
    uint32_t abcd[4];
    uint32_t sse4_1_mask = (1 << 19);

    run_cpuid(1, 0, abcd);
    /* CPUID.(EAX=01H, ECX=0H):ECX.SSE41[bit 19]==1 */
    return (abcd[2] & sse4_1_mask) != 0;
}


/**
 * Checks whether current CPU supports SSE2 instruction set
 *
 * Source:
 *     http://stackoverflow.com/questions/6121792/how-to-check-if-a-cpu-supports-the-sse3-instruction-set
 *     http://stackoverflow.com/a/7495023
 */
int check_sse2()
{
    uint32_t abcd[4];
    uint32_t sse2_mask = (1 << 26);

    run_cpuid(1, 0, abcd);
    /* CPUID.(EAX=01H, ECX=0H):EDX.SSE2[bit 26]==1 */
    return (abcd[3] & sse2_mask) != 0;
}

#endif /* non-Intel compiler */


/**
 * Checks whether current CPU supports AVX2 and other New Haswell features
 */
bool can_use_intel_core_4th_gen_features()
{
    static int the_4th_gen_features_available = -1;
    /* test is performed once */
    if (the_4th_gen_features_available < 0 )
        the_4th_gen_features_available = check_4th_gen_intel_core_features();

    return the_4th_gen_features_available;
}


/**
 * Checks whether current CPU supports SSE4.1 instruction set
 */
bool can_use_sse4_1()
{
    return check_sse4_1();
}


/**
 * Checks whether current CPU supports SSE2 instruction set
 */
bool can_use_sse2()
{
    return check_sse2();
}
