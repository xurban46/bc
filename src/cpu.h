/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


/*
    CPU new instruction set check.
    Source: https://software.intel.com/en-us/articles/how-to-detect-new-instruction-support-in-the-4th-generation-intel-core-processor-family
 */


#pragma once

#include <stdbool.h>
#include <cpuid.h>


#define SIMD_PADDING_BYTES 32


/**
 * Checks whether current CPU supports AVX2 and other New Haswell features
 */
bool can_use_intel_core_4th_gen_features();


/**
 * Checks whether current CPU supports SSE4.1 instruction set
 */
bool can_use_sse4_1();


/**
 * Checks whether current CPU supports SSE2 instruction set
 */
bool can_use_sse2();



static inline bool can_use_simd() {
    #ifdef AVX2
        if (can_use_intel_core_4th_gen_features()) {
            return true;
        }
    #endif

    #ifdef SSE2
        if (can_use_sse2()) {
            return true;
        }
    #endif

    return false;
}
