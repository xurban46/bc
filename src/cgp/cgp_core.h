/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include "../ga.h"
#include "cgp_config.h"


#define CGP_FUNC_INPUTS 2
#define CGP_NODES (CGP_COLS * CGP_ROWS)
#define CGP_CHR_OUTPUTS_INDEX ((CGP_FUNC_INPUTS + 1) * CGP_NODES)
#define CGP_CHR_LENGTH (CGP_CHR_OUTPUTS_INDEX + CGP_OUTPUTS)

static const ga_problem_type_t CGP_PROBLEM_TYPE = maximize;

typedef unsigned char cgp_value_t;

#define CGP_FUNC_COUNT 16
typedef enum
{
    c255 = 0,   // 255
    identity,   // a
    inversion,  // 255 - a
    b_or,       // a or b
    b_not1or2,  // (not a) or b
    b_and,      // a and b
    b_nand,     // not (a and b)
    b_xor,      // a xor b
    rshift1,    // a >> 1
    rshift2,    // a >> 2
    swap,       // a <-> b
    add,        // a + b
    add_sat,    // a +S b
    avg,        // (a + b) >> 1
    max,        // max(a, b)
    min,        // min(a, b)
} cgp_func_t;


/**
 * One CGP node (function block)
 */
typedef struct {
    int inputs[CGP_FUNC_INPUTS];
    cgp_func_t function;
    bool is_active;
} cgp_node_t;


/**
 * Chromosome
 */
struct cgp_genome {
    cgp_node_t nodes[CGP_COLS * CGP_ROWS];
    int outputs[CGP_OUTPUTS];
};
typedef struct cgp_genome* cgp_genome_t;


/**
 * Initialize CGP internals
 */
void cgp_init(int mutation_rate, ga_fitness_func_t fitness_func);


/**
 * Deinitialize CGP internals
 */
void cgp_deinit();


/**
 * Create a new CGP population with given size
 * @param  mutation rate (in number of genes)
 * @return
 */
ga_pop_t cgp_init_pop(int pop_size);


/**
 * Allocates memory for new CGP genome
 * @return pointer to newly allocated genome
 */
void* cgp_alloc_genome();


/**
 * Initializes CGP genome to random values
 * @param chromosome
 * @return 0 on success, other value on error
 */
int cgp_randomize_genome(ga_chr_t chromosome);


/**
 * Deinitialize CGP genome
 * @param  genome
 * @return
 */
void cgp_free_genome(void *genome);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_copy_genome(void *_dst, void *_src);


/**
 * Replace gene on given locus with random alele
 * @param chr
 * @param gene
 * @return whether active node was changed or not (phenotype has changed)
 */
bool cgp_randomize_gene(cgp_genome_t genome, int gene);


/**
 * Mutate given chromosome
 * @param chr
 */
void cgp_mutate_chr(ga_chr_t chromosome);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_replace_chr(ga_chr_t chromosome, ga_chr_t replacement);


/**
 * Calculate output of given chromosome and inputs
 * @param chr
 */
void cgp_get_output(ga_chr_t chromosome, cgp_value_t *inputs, cgp_value_t *outputs);


/**
 * Create new generation
 * @param pop
 * @param mutation_rate
 */
void cgp_offspring(ga_pop_t pop);


/**
 * Returns index of node in given column and row
 * @param  col
 * @param  row
 * @return
 */
static inline int cgp_node_index(int col, int row)
{
    return CGP_ROWS * col + row;
}


/**
 * Returns node column based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_col(int index)
{
    return index / CGP_ROWS;
}


/**
 * Returns node row based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_row(int index)
{
    return index % CGP_ROWS;
}


/**
 * Finds which blocks are active.
 * @param chromosome
 * @param active
 */
void cgp_find_active_blocks(ga_chr_t chromosome);
