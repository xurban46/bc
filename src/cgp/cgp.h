/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include "cgp_core.h"
#include "cgp_dump.h"
#include "cgp_load.h"
