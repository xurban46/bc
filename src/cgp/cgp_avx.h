/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include <immintrin.h>

#include "cgp_core.h"


typedef __m256i __m256i_aligned __attribute__ ((aligned (32)));


/**
 * Calculate output of given chromosome and inputs using AVX instructions
 * @param chr
 * @param inputs
 * @param outputs
 */
void cgp_get_output_avx(ga_chr_t chromosome, __m256i_aligned inputs[CGP_INPUTS], __m256i_aligned outputs[CGP_OUTPUTS]);
