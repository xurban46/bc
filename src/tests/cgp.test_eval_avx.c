/**
 * Tests CGP evaluation = calculation of the outputs.
 * Compile with -DTEST_EVAL_AVX -DAVX2
 * Source files cgp_core.c cgp_dump.c cgp_avx.c cpu.c ga.c
 */

#include <stdlib.h>
#include <stdio.h>
#include <immintrin.h>

#include "../cpu.h"
#include "../cgp.h"
#include "../cgp_avx.h"



#define in1 {0, 1, 2, 3, 4, 5, 6, 7, 8}
#define in4 in1, in1, in1, in1
#define in16 in4, in4, in4, in4
#define in32 in16, in16

#define rep4(x) (x), (x), (x), (x)
#define rep16(x) rep4((x)), rep4((x)), rep4((x)), rep4((x))
#define rep32(x) rep16((x)), rep16((x))


int main(int argc, char const *argv[])
{
    // pre-flight check
    if (!can_use_intel_core_4th_gen_features()) {
        fprintf(stderr, "%s", "AVX2 not supported.\n");
        exit(1);
    }


    unsigned char _inputs[CGP_INPUTS][32] = {
        {rep32(0)},
        {rep32(1)},
        {rep32(2)},
        {rep32(3)},
        {rep32(4)},
        {rep32(5)},
        {rep32(6)},
        {rep32(7)},
        {rep32(8)},
    };

    __m256i_aligned inputs[CGP_INPUTS];
    __m256i_aligned outputs[CGP_OUTPUTS];

    for (int i = 0; i < CGP_INPUTS; i++) {
        inputs[i] = _mm256_load_si256((__m256i*)(&_inputs[i]));
    };

    cgp_init(0, NULL);

    cgp_genome_t genome = (cgp_genome_t) cgp_alloc_genome();
    struct ga_chr chr = {
        .genome = genome
    };

    // define nodes
    for (int x = 0; x < CGP_COLS; x++) {
        for (int y = 0; y < CGP_ROWS; y++) {
            int i = cgp_node_index(x, y);
            cgp_node_t *n = &(genome->nodes[i]);
            n->inputs[0] = i + CGP_INPUTS - y - y - 1;
            n->inputs[1] = i + CGP_INPUTS - CGP_ROWS;
            n->function = (cgp_func_t) (i % CGP_FUNC_COUNT);
        }
    }

    // define outputs
    genome->outputs[0] = 13;

    cgp_dump_chr_asciiart(&chr, stdout, false);
    putchar('\n');
    cgp_get_output_avx(&chr, inputs, outputs);

    free(genome);
    cgp_deinit();
}
