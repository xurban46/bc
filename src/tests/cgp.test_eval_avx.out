Inputs: 9
Outputs: 1
Size: 8 x 4
Blocks: 2-ary, 1 output(s), 16 functions
Fitness: <none>
     .------------------------------------------------------------------------------------------------------------------------------------------------.
     |      .----.            .----.            .----.            .----.            .----.            .----.            .----.            .----.      |
[ 0]>| [ 8]>|    |>[ 9]  [12]>|    |>[13]  [16]>|    |>[17]  [20]>|    |>[21]  [24]>|    |>[25]  [28]>|    |>[29]  [32]>|    |>[33]  [36]>|    |>[37] |>[13]
[ 1]>| [ 5]>| FF |       [ 9]>|~1|2|       [13]>|a>>1|       [17]>| +S |       [21]>| FF |       [25]>|~1|2|       [29]>|a>>1|       [33]>| +S |      |
[ 2]>|      '----'            '----'            '----'            '----'            '----'            '----'            '----'            '----'      |
[ 3]>|      .----.            .----.            .----.            .----.            .----.            .----.            .----.            .----.      |
[ 4]>| [ 7]>|    |>[10]  [11]>|    |>[14]  [15]>|    |>[18]  [19]>|    |>[22]  [23]>|    |>[26]  [27]>|    |>[30]  [31]>|    |>[34]  [35]>|    |>[38] |
[ 5]>| [ 6]>|  a |       [10]>| and|       [14]>|a>>2|       [18]>| avg|       [22]>|  a |       [26]>| and|       [30]>|a>>2|       [34]>| avg|      |
[ 6]>|      '----'            '----'            '----'            '----'            '----'            '----'            '----'            '----'      |
[ 7]>|      .----.            .----.            .----.            .----.            .----.            .----.            .----.            .----.      |
[ 8]>| [ 6]>|    |>[11]  [10]>|    |>[15]  [14]>|    |>[19]  [18]>|    |>[23]  [22]>|    |>[27]  [26]>|    |>[31]  [30]>|    |>[35]  [34]>|    |>[39] |
     | [ 7]>|FF-a|       [11]>|nand|       [15]>|swap|       [19]>| max|       [23]>|FF-a|       [27]>|nand|       [31]>|swap|       [35]>| max|      |
     |      '----'            '----'            '----'            '----'            '----'            '----'            '----'            '----'      |
     |      .----.            .----.            .----.            .----.            .----.            .----.            .----.            .----.      |
     | [ 5]>|    |>[12]  [ 9]>|    |>[16]  [13]>|    |>[20]  [17]>|    |>[24]  [21]>|    |>[28]  [25]>|    |>[32]  [29]>|    |>[36]  [33]>|    |>[40] |
     | [ 8]>| or |       [12]>| xor|       [16]>| +  |       [20]>| min|       [24]>| or |       [28]>| xor|       [32]>| +  |       [36]>| min|      |
     |      '----'            '----'            '----'            '----'            '----'            '----'            '----'            '----'      |
     '------------------------------------------------------------------------------------------------------------------------------------------------'

I:  0 = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
I:  1 = 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
I:  2 = 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2
I:  3 = 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3
I:  4 = 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4
I:  5 = 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5
I:  6 = 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6
I:  7 = 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7
I:  8 = 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8
N:  9 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
N: 10 = 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7
N: 11 = 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249, 249
N: 12 = 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13
N: 13 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
N: 14 = 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
N: 15 = 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254, 254
N: 16 = 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242, 242
N: 17 = 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121
N: 18 = 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63
N: 19 = 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30
N: 20 = 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241, 241
N: 21 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
N: 22 = 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46
N: 23 = 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63
N: 24 = 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121, 121
N: 25 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
N: 26 = 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63
N: 27 = 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209, 209
N: 28 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
N: 29 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
N: 30 = 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17
N: 31 = 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238, 238
N: 32 = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
N: 33 = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
N: 34 = 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59
N: 35 = 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30
N: 36 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
N: 37 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
N: 38 = 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44, 44
N: 39 = 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59, 59
N: 40 = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
O:  0 = 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255
